### DISCUSSION TITLE
Add pure-python functions to the pybind11 library

### DISCUSSION THREAD

**Discussion by a-jp** [pybind/pybind11#5047] [Q&A]
Hi,
So far my pybind11 library embeds the interpreter (pybind11_add_module) and produces me a library of wrapped c++ code, that I can call from python.  To allow seamless operation for the user, I have functions that would be better/faster/easier to write in pure python, rather than writing them in long-hand in c++ and creating a wrapper. But, I'd like them to be "added" (I know that's the wrong word) to the same python module that pybind11 creates for my pure c++ code.  Is that possible so that the user could call them and in this instance they happen (unknowingly) to be just calling pure python from another .py file?  Could I have a MWE to do that?  Hopefully I've explained that correctly. Thanks

**Comment by a-jp** [comment]
Could anyone help here? Even just to confirm that the above is or is not possible with pybind11?

**Comment by jnooree** [comment]
You could just import the file and/or function from C++ side.

Ref: https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#accessing-python-libraries-from-c

**Reply by a-jp** [reply]
Ok yep, but how to I add that to the python ".so" that is generated by my cmake that I can then import into other python file.  I'd like the pure-python function added to that singular python library that at present just exposes my c++ code via pybind11.

**Reply by jnooree** [reply]
That would be impossible I guess then.. Maybe publish the python files in pypi then instruct the users to install it?

**Reply by jnooree** [reply]
Oh, using py::eval/py::exec might do the trick (though quite a dirty hack IMO)

https://pybind11.readthedocs.io/en/stable/advanced/pycpp/utilities.html
